src/contexts/NetworkInfo.js

```diff
@@ -42,24 +62,81 @@ export default function Provider({ children }) {
     })
   }, [])
 
+  const updateTokensList = useCallback((chainId, tokensList) => {
+    dispatch({
+      type: UPDATE_TOKENS_LIST,
+      payload: {
+        chainId,
+        tokensList,
+      },
+    })
+  }, [])
+
   return (
     <NetworkInfoContext.Provider
       value={useMemo(
         () => [
           state,
           {
             updateChain,
+            updateTokensList,
           },
         ],
-        [state, updateChain]
+        [state, updateChain, updateTokensList]
       )}
     >
       {children}
     </NetworkInfoContext.Provider>
   )
 }
 
+async function getTokenList(listUrl) {
+  let response
+  try {
+    response = await fetch(listUrl)
+  } catch (error) {
+    console.debug('Failed to fetch list', listUrl, error)
+    throw new Error(`Failed to download list ${listUrl}`)
+  }
+
+  if (!response.ok) {
+    throw new Error(`Failed to download list ${listUrl}`)
+  }
+
+  const json = await response.json()
+  // if (!tokenListValidator(json)) {
+  //   const validationErrors: string =
+  //     tokenListValidator.errors?.reduce<string>((memo, error) => {
+  //       const add = `${error.dataPath} ${error.message ?? ''}`
+  //       return memo.length > 0 ? `${memo}; ${add}` : `${add}`
+  //     }, '') ?? 'unknown error'
+  //   throw new Error(`Token list failed validation: ${validationErrors}`)
+  // }
+  return json
+}
+
 export function useNetworksInfo() {
   const [{ networksInfo }, { updateChain }] = useNetworksInfoContext()
   return [networksInfo.filter(Boolean).length ? networksInfo : [NetworksInfoEnv[0]], updateChain]
 }
+
+export function useTokensList() {
+  const [networksInfo] = useNetworksInfo()
+  const [{ tokensList }, { updateTokensList }] = useNetworksInfoContext()
+
+  useEffect(() => {
+    const fetchTokensList = async index => {
+      try {
+        const tokenList = await getTokenList(networksInfo[index].tokensListUrl)
+        updateTokensList(networksInfo[index].chainId, tokenList)
+      } catch (e) {}
+    }
+    networksInfo.forEach((networkInfo, index) => {
```

Comment:
please use Promise.all

Reply:
We dont await for it's result, so we dont need to use Promise.all here

Reply:
memoRequest is an async function. If you user forEach here, it will run sequentially

Reply:
yea, that's my intended

Reply:
Wait, my bad for misunderstanding, forEach here would not making it run sequentially. It would call api async together

Reply:
Added some console.log to demo this
![image](https://user-images.githubusercontent.com/23650671/165012974-ba2b7353-0c01-4edf-b55a-231c5cd803c3.png)
<img width="697" alt="image" src="https://user-images.githubusercontent.com/23650671/165012998-03c08804-f227-4b04-a37a-7efcf3dbc455.png">